### PR TITLE
out variant for cudnn batch norm

### PR DESCRIPTION
**Pull request by makslevental**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61824
* #61821
* #61802
* #61801
* #61800

Differential Revision: [D29812546](https://our.internmc.facebook.com/intern/diff/D29812546)

### CODE REVIEW

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812546).

File: aten/src/ATen/native/cudnn/BatchNorm.cpp [hunk 1]
```diff
@@ -195,41 +201,103 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> cudnn_batch_norm(
...
+  Tensor save_mean, save_var, reserve;
+  cudnnBatchNormMode_t mode = getCudnnBatchNormMode(
+      training, input->suggest_memory_format(), input->dim());
+
+  if (training) {
+    int64_t num_features = input_t.size(1);
+    save_mean = at::empty({num_features}, weight_t.options());
+    save_var = at::empty({num_features}, weight_t.options());
+
+#if CUDNN_VERSION >= 7400
```
Comment (desertfire) [inline aten/src/ATen/native/cudnn/BatchNorm.cpp:273]: `reserve` is not allocated for `#else` or `!training`?

File: aten/src/ATen/native/cudnn/BatchNorm.cpp [hunk 2]
```diff
@@ -195,41 +201,103 @@ std::tuple<Tensor, Tensor, Tensor, Tensor> cudnn_batch_norm(
...
+    size_t workspace_size;
+    // get the reserved size and allocate as tensor
+    size_t reserve_size;
+    AT_CUDNN_CHECK(cudnnGetBatchNormalizationTrainingExReserveSpaceSize(
+        handle, mode, op, nullptr, idesc.desc(), &reserve_size));
+    reserve = at::empty(reserve_size, input->options().dtype(kByte));
+#endif // CUDNN_VERSION >= 7400
+  }
+
+  std::tie(output_t, save_mean, save_var, reserve) = cudnn_batch_norm_out(
```
Comment (desertfire) [inline aten/src/ATen/native/cudnn/BatchNorm.cpp:284]: Would linter complain if we skip the assignment to std::tie?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -1309,6 +1309,10 @@
   dispatch:
     CUDA: cudnn_batch_norm
 
+- func: cudnn_batch_norm.out(Tensor input, Tensor weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float exponential_average_factor, float epsilon, *, Tensor(a!) out, Tensor(b!) mean, Tensor(c!) var, Tensor(d!) reserve) -> (Tensor(a!), Tensor(b!), Tensor(c!), Tensor(d!))
```
Comment (desertfire) [inline aten/src/ATen/native/native_functions.yaml:1312]: Nit: keep the new parameter naming consistent? e.g. `save_mean`, `save_var`

File: aten/src/ATen/native/cudnn/BatchNorm.cpp (hunk 1 above)
Comment (makslevental) [reply aten/src/ATen/native/cudnn/BatchNorm.cpp:273]: my reasoning here was three fold:

1) a reserve buffer is indeed not used, neither for cudnn < 7.4 nor in forward pass  
2) up above we have this

>  // save_mean and save_var can be undefined
  // If this causes problems, we can initialize them to empty tensors
  // of the correct type
  
 thus reasoning by analogy...

3) return is ultimately consumed in [Tensor batch_norm](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp#L538)
```
return std::get<0>(at::_batch_norm_impl_index(input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled));
```
which discards. 

but it was just my intuition 🤷. if you think this would break something i will certainly put it back.

File: aten/src/ATen/native/cudnn/BatchNorm.cpp (hunk 2 above)
Comment (makslevental) [reply aten/src/ATen/native/cudnn/BatchNorm.cpp:284]: C++ doesn't do this kind of unpacking before [`std=c++17`](https://en.cppreference.com/w/cpp/language/structured_binding) and our cmake sets [`std=c++14`](https://github.com/pytorch/pytorch/blob/master/CMakeLists.txt#L40)

Comment (makslevental) [reply aten/src/ATen/native/cudnn/BatchNorm.cpp:284]: oh duh whoops. will change. thanks!

**Review by desertfire — APPROVED** [review]

**Review by desertfire — APPROVED** [review]

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812546).

**Comment by makslevental** [comment]
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29812546).

**Comment by vadimkantorov** [comment]
This does not modify for now the user-facing nn.BatchNorm*d and F.batch_norm, right?

**Comment by vadimkantorov** [comment]
@ezyang if the CLA is signed, maybe staleness could be reevaluated?
